### PR TITLE
Remove `pagy_cursor`

### PR DESCRIPTION
**Pull request by jagthedrummer**
This is the simplest thing I could get to work for removing `pagy_cursor`. Luckily we seem to only use it for one feature, which is an `after` filter. That filter only seems to be used by API clients, and is not used via the UI (we use the standard `page` for `pagy` there). I don't personally use the API in BT for anything, so I'm not sure if there are subtleties in there that I'm missing.

I previously added a test for pagination here: https://github.com/bullet-train-co/bullet_train-core/pull/1131/files

### CODE REVIEW

**Comment by newstler** [comment]
@jagthedrummer Not sure what's the point of replacing `pagy`'s native method with a custom one and leaving the rest of the gem do it's work. Also, haven't seen the code for a while, so would need to dig a bit deeper in case it's needed. But left my comments, and also feels like it's possible to reuse `apply_pagination` logic in `collection_has_more?` to avoid duplications.

**Comment by jagthedrummer** [comment]
@newstler, this isn't replacing a `pagy` native method, it's replacing the one single method from `pagy_cursor` that we use, and will allow us to drop `pagy_cursor` as a dependency. Currently `pagy_cursor` is keeping us pinned to `pagy` version 8, and is preventing us from upgrading to version 9. Maybe I should go ahead and remove that dependency as part of this PR.

Good call on consolidating some of that logic into a single spot. I'll clean that up tomorrow.

**Comment by jagthedrummer** [comment]
OK, @newstler, I removed `pagy_cursor` completely and refactored `Api::Controllers::Base` a bit to reduce duplication.

**Review by RichStone — CHANGES_REQUESTED** [review]
Thanks for taking a stab at this, it's actually a cool idea for getting off `pagy_cursor`.

At the same time, it feels a bit risky moving off a proven system so I had a thorough look at this and I can't find logical or performance issues at this point. 😅
The change request is mainly to add tests. If we are going to hand-roll our own cursor-based pagination, we should test different corner cases (last ID in collection, wrong IDs passed, ID 10 passed in a total collection of 20...). One "e2e" happy path test scenario could look like this:

```ruby
test "index pagination defaults to ascending by ID and handles reverse order" do
  pagy_items_limit = 20
  # Creating just enough records to have 2 pages.
  teams = create_list(:team, pagy_items_limit + 1)

  # Test default pagination order which is ascending by ID.
  get "/api/v2/teams", params: {access_token: access_token}
  parsed_body = response.parsed_body
  assert_equal 20, parsed_body.count
  assert_equal teams.first.id, parsed_body.first["id"]

  # :after param provides records with higher IDs.
  get "/api/v2/teams",
    params: {access_token: access_token, after: parsed_body.last["id"]}
  parsed_body = response.parsed_body
  assert_equal 1, parsed_body.count
  assert_equal all_teams.last.id, parsed_body.first["id"]

  # order: desc reverses collection order.
  get "/api/v2/teams", params: {access_token: access_token, sort_order: :desc}
  parsed_body = response.parsed_body
  assert_equal 20, parsed_body.count
  assert_equal all_teams.last.id, parsed_body.first["id"]

  # :after params now returns records that come on the next page with lower IDs.
  get "/api/v2/teams",
    params: {access_token: access_token, sort_order: :desc, after: parsed_body.last["id"]}
  parsed_body = response.parsed_body
  assert_equal 1, parsed_body.count
  assert_equal all_teams.first.id, parsed_body.first["id"]
end
```

Let me know if you think that makes sense or need some support with it!

⚠️ Probably worth mentioning in the release notes: This will break most downstreams that have apply_pagination overridden (which I think most would have who offer an API, since our historical defaults are not as useful - like ascending order, no filtering...).

cc @Riddlerrr to please also take a look at the changes there. 🙏

**Comment by RichStone** [comment]
Oh, I see now that you've added some tests here: #1131
Then we are off to a good start!

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb [hunk 1]
```diff
@@ -30,18 +27,24 @@ def modify_url_params(url, new_params)
...
+        link_header = response.headers["Link"]
+        link_value = "<#{modify_url_params(request.url, after: last_id_in_collection)}>; rel=\"next\""
+        response.headers["Link"] = link_header ? "#{link_header}, #{link_value}" : link_value
+        response.headers["Pagination-Next"] = last_id_in_collection
       end
     end
 
+    def collection_has_more?
+      return false unless last_id_in_collection
+      remaining_collection = collection.limit(nil).order(id: :asc).where("id > ?", last_id_in_collection)
```
[the remark this answers is not in the document]
Comment (jagthedrummer) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:40]: Yes.

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb [hunk 2]
```diff
@@ -97,10 +100,22 @@ def collection_variable
...
+  def collection=(new_collection)
+    @collection = new_collection
+    instance_variable_set collection_variable, new_collection
+  end
+
   def apply_pagination
-    collection = instance_variable_get collection_variable
-    @pagy, collection = pagy_cursor collection, after: params[:after], order: {id: :asc}
-    instance_variable_set collection_variable, collection
+    pagination_collection = collection.order(id: :asc)
```
[the remark this answers is not in the document]
Comment (jagthedrummer) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:113]: Also, I don't think we can change change the sort order without also changing how we paginate. Changing the order changes the meaning of the `after` param.

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb [hunk 3]
```diff
@@ -30,18 +27,24 @@ def modify_url_params(url, new_params)
...
     end
 
+    def collection_has_more?
+      return false unless last_id_in_collection
+      remaining_collection = collection.limit(nil).order(id: :asc).where("id > ?", last_id_in_collection)
+      remaining_collection.any?
+    end
+
+    def last_id_in_collection
+      @last_id_in_collection ||= collection&.last&.id
```
Comment (Riddlerrr) [inline bullet_train-api/app/controllers/concerns/api/controllers/base.rb:45]:
```suggestion
      @last_id_in_collection ||= collection&.maximum(:id)
````

Small improvement that helps to load less from DB here. It sends the `SELECT MAX("model"."id") ...` query instead of loading the whole model instance and using only the id from there.

**Review by Riddlerrr — APPROVED** [review]
Amazing! I like this small, elegant solution that eliminates the need for the gem dependency altogether.

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb (hunk 3 above)
Comment (jagthedrummer) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:45]: @Riddlerrr  I think this would result in two queries instead of one. We're always going to have to load the entire collection, so there's really no benefit to only temporarily avoiding it, especially if the cost is an extra query.

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (RichStone) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:40]: That was by mistake.

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (RichStone) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:113]: > Also, I don't think we can change change the sort order without also changing how we paginate. Changing the order changes the meaning of the after param.

Yes, it becomes before then...

> I believe that this is preserving what we already did before this PR. If we need to change the default that should be a different PR and we should be loud about it.

Same logic applies here: https://github.com/bullet-train-co/bullet_train-core/pull/1130#discussion_r2109319872
(hence my questions in the comment there)

**Review by RichStone — APPROVED** [review]
> @RichStone Is there something else you're wanting on this PR? (You're currently blocking it from being merged until you re-review.)

Sorry for the delay! Gonna unblock, but I think the existing test could show a bit better the first and last records especially when going to the next pages.

> We're always going to have to load the entire collection, so there's really no benefit to only temporarily avoiding it, especially if the cost is an extra query.

I hope it's never the "entire" entire collection. 😅

File: bullet_train-api/app/controllers/concerns/api/controllers/base.rb [hunk 4]
```diff
@@ -97,10 +100,22 @@ def collection_variable
...
+
   def apply_pagination
-    collection = instance_variable_get collection_variable
-    @pagy, collection = pagy_cursor collection, after: params[:after], order: {id: :asc}
-    instance_variable_set collection_variable, collection
+    pagination_collection = collection.order(id: :asc)
+    if params[:after]
+      pagination_collection = pagination_collection.where("id > ?", params[:after])
+    end
+    @pagy, pagination_collection = pagy(pagination_collection)
```
[the remark this answers is not in the document]
Comment (jagthedrummer) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:117]: @RichStone How do you figure that "we currently offer an API contract of 20 records per page"? Is there somewhere that we currently actively specify 20 records per page? I'm not finding anything like that...

[the remark this answers is not in the document]
Comment (jagthedrummer) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:117]: @RichStone to my mind that is not a "contract" in any way shape or form. We (as in the BT maintainers / the project as a whole) have never made any kind of a promise or even so much as a statement suggesting that we will always provide you with 20 items per page and that we will never change it or allow it to be changed. You seem to have made an assumption along those lines. If you want that number to never change in _your_ project I think you should modify the `pagy` initializer to set it explicitly.

[the remark this answers is not in the document]
Comment (RichStone) [reply bullet_train-api/app/controllers/concerns/api/controllers/base.rb:117]: @jagthedrummer, I think we both agree that we generally try to avoid backward-incompatible changes for BT users and also for the API clients of our BT users.

One interface/feature that we offer for the public API is pagination, that's what I called "contract". I guess it's some public web API speech, anything you offer publicly is a contract/promise, be it written down or not, you have to assume that it's being used by API clients.

Different from UI users, the change of the default number of records on a page can be a breaking change for some API clients, depending on how they serialize the data. Hence my suggestion to be explicit about our current implicit promise, to avoid inadvertent changes in the future. :)
